stan/math/rev/core/arena_matrix.hpp

```diff
@@ -139,6 +135,264 @@ class arena_matrix : public Eigen::Map<MatrixType> {
+   * @param other matrix to copy from
+   */
+  arena_matrix(arena_matrix<MatrixType>&& other)
+      : Base::Map(other.rows(), other.cols(), other.nonZeros(),
+                  other.outerIndexPtr(), other.innerIndexPtr(),
+                  other.valuePtr()) {}
+  /**
+   * Copy constructor. No actual copy is performed
+   * @note Since the memory for the arena matrix sits in Stan's memory arena all
+   * copies/moves of arena matrices are shallow moves
+   * @param other matrix to copy from
+   */
+  arena_matrix(arena_matrix<MatrixType>& other)
+      : Base::Map(other.rows(), other.cols(), other.nonZeros(),
+                  other.outerIndexPtr(), other.innerIndexPtr(),
+                  other.valuePtr()) {}
+
+  // without this using, compiler prefers combination of implicit construction
+  // and copy assignment to the inherited operator when assigned an expression
+  using Base::operator=;
+
+  /**
+   * Copy assignment operator.
+   * @tparam ArenaMatrix An `arena_matrix` type
+   * @param other matrix to copy from
+   * @return `*this`
+   */
+  template <typename ArenaMatrix,
+            require_same_t<ArenaMatrix, arena_matrix<MatrixType>>* = nullptr>
+  arena_matrix& operator=(ArenaMatrix&& other) {
+    // placement new changes what data map points to - there is no allocation
+    new (this) Base(other.rows(), other.cols(), other.nonZeros(),
+                    const_cast<StorageIndex*>(other.outerIndexPtr()),
+                    const_cast<StorageIndex*>(other.innerIndexPtr()),
+                    const_cast<Scalar*>(other.valuePtr()));
+    return *this;
+  }
+
+  /**
+   * Assignment operator for assigning an expression.
+   * @tparam Expr An expression that an `arena_matrix<MatrixType>` can be
+   * constructed from
+   * @param expr expression to evaluate into this
+   * @return `*this`
+   */
+  template <typename Expr,
+            require_not_same_t<Expr, arena_matrix<MatrixType>>* = nullptr>
+  arena_matrix& operator=(Expr&& expr) {
+    *this = arena_matrix(std::forward<Expr>(expr));
+    return *this;
+  }
+
+ private:
+  /**
+   * inplace operations functor for `Sparse (.*)= Sparse`.
+   * @note This assumes that each matrix is of the same size and sparsity
+   * pattern.
+   * @tparam F A type with a valid `operator()(Scalar& x, const Scalar& y)`
+   * method
+   * @tparam Expr Type derived from `Eigen::SparseMatrixBase`
+   * @param f Functor that performs the inplace operation
+   * @param other The right hand side of the inplace operation
+   */
+  template <typename F, typename Expr,
+            require_convertible_t<Expr&, MatrixType>* = nullptr>
+  void inplace_ops_impl(F&& f, Expr&& other) {
+    auto&& x = to_ref(other);
+    for (int k = 0; k < (*this).outerSize(); ++k) {
+      typename Base::InnerIterator it(*this, k);
+      typename std::decay_t<Expr>::InnerIterator iz(x, k);
+      for (; static_cast<bool>(it) && static_cast<bool>(iz); ++it, ++iz) {
+        f(it.valueRef(), iz.value());
+      }
```

Comment:
Should this (and the other overloads) just loop over `(*this).innerSize()`? The compiler might be able to optimise better if it knows the length.

Otherwise, I think this is more readable as a `while` loop:
```cpp
while (static_cast<bool>(it) && static_cast<bool>(iz)) {
  f(it.valueRef(), iz.value());
  ++it;
  ++iz;
}
```

Reply:
Yes! So actually since we know for the sparse arena matrix the pointer for the values is contiguous from our memory arena we can actually just loop over the value pointer itself with the number of nonzero entries. At least for the scalar and arena matrix inplace ops. For general sparse and dense idt we have that guarantee so we should still do something like the while loop you have in the above